File: 7/min_dist_between_bst_nodes.go
package main

import (
	"fmt"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func min(a, b int) int {
	if a <= b {
		return a
	}
	return b
}

func inOrder(root *TreeNode, previousVal, result *int) {
	if root.Left != nil {
		inOrder(root.Left, previousVal, result)
	}
	*result = min(*result, root.Val-*previousVal)
	*previousVal = root.Val
	if root.Right != nil {
		inOrder(root.Right, previousVal, result)
	}
}

func minDiffInBST(root *TreeNode) int {
	previousVal := -2147483647
	result := 2147483647
	inOrder(root, &previousVal, &result)
	return result
}

func main() {
	root := TreeNode{4, &TreeNode{2, &TreeNode{1, nil, nil}, &TreeNode{3, nil, nil}}, &TreeNode{6, nil, nil}}
	fmt.Println(minDiffInBST(&root))
}

File: 7/min_dist_between_bst_nodes.py
import sys
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    previous_val = -sys.maxsize
    result = sys.maxsize

    def minDiffInBST(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root.left:
            self.minDiffInBST(root.left)
        self.result = min(self.result, root.val-self.previous_val)
        self.previous_val = root.val
        if root.right:
            self.minDiffInBST(root.right)


        return self.result
        
